File: week-06/ek_7576_tomato.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <cmath>
using namespace std;

int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};
int n,m;
vector<vector<int>> tomato;

int FindAnswer() {
    int ans = -1;
    for(int row = 0; row < n; row++) {
        for (int col = 0; col < m; col++) {
            if(tomato[row][col] == 0) {
                return -1;
            }
            else if(tomato[row][col] > 0) {
                ans = max(ans, tomato[row][col]);
            }
        }
    }
    return ans-1;
}

int main(void) {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> m >> n;
    tomato.assign(n, vector<int> (m));

    queue<pair<int,int>> q;
    for (int row = 0; row < n; row++) {
        for (int col = 0; col < m; col++) {
            cin >> tomato[row][col];

            if(tomato[row][col] == 1)
                q.push({row, col});
        }
    }

    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for(int k = 0; k < 4; k++) {
            int nx = x + dx[k];
            int ny = y + dy[k];

            if (0 <= nx && nx < n && 0 <= ny && ny < m && tomato[nx][ny] == 0) {
                tomato[nx][ny] = tomato[x][y] + 1;
                q.push({nx,ny});
            }
        }
    }

    cout << FindAnswer();
    
    return 0;
}

File: week-06/ek_7576_tomato.py
import sys
from collections import deque
read = sys.stdin.readline

dx = (1,0,-1,0)
dy = (0,1,0,-1)

def bfs():
    while dq:
        x, y = dq.popleft()

        for k in range(4):
            nx, ny = x + dx[k], y + dy[k]

            # 범위 안에 있음, 안 익은 토마토, 방문 안했음
            if 0 <= nx < n and 0 <= ny < m and tomato[nx][ny] == 0:
                dq.append((nx,ny))
                tomato[nx][ny] = tomato[x][y] + 1

def find_answer() -> int:
    ans = -1

    for row in range(n):
        for col in range(m):
            # 토마토 익었음
            if tomato[row][col] > 0:
                ans = max(ans, tomato[row][col])
            # 안 익은 토마토가 있음
            elif tomato[row][col] == 0:
                return -1
            
    return ans - 1

# 입력 받기
m, n = map(int, read().split())
tomato = [list(map(int, read().split())) for _ in range(n)]

dq = deque()

# 1. 익은 토마토 찾기
for row in range(n):
    for col in range(m):
        if tomato[row][col] == 1:
            dq.append((row, col))

# 2 토마토 익히기
bfs()
# 3 최소 날짜 찾기
print(find_answer())